File: js/minimax_creator/sampling.js
// The sampler row, drawn the same way on both nodes.
//
// Both the Creator and the Timeline own their sampler and declare the same
// widgets under the same names, so there is one row and both mount it. It lives
// outside the panel on either node because the panel says what the piece *is*
// and this says how it is run.
//
// The widgets are the real ComfyUI ones, hidden by the entry point and re-drawn
// here: `graphToPrompt` reads values off `node.widgets`, so these pills write
// through to them rather than holding state of their own.

import { el, icon } from "./dom.js";
import { t } from "./i18n.js";
import { openChoicePopover, stepperPill } from "./pills.js";
import { uiSetting } from "./api.js";
import { lastSeed } from "./seedmemory.js";

export const SEED_CONTROL = ["fixed", "increment", "decrement", "randomize"];

// Every widget this row draws. The entry point hides exactly these, so a name
// added here without being added there would render twice — once as a pill and
// once as the stock widget underneath.
export const SAMPLING_WIDGETS = [
  "seed", "control_after_generate", "steps", "cfg", "sampler_name", "scheduler",
  "shift_video", "shift_audio",
  "block_cache", "spectrum", "spectrum_blend", "sage", "attention", "chunk_ffn",
  "fp16_accumulation",
];

// `sage` is in that list and is never drawn: it is the switch `attention`
// replaced, kept so a workflow saved with it on still runs sage, and hidden so
// nobody sets it from two places. See `adoptSage`.

const ATTENTION_TITLE = {
  default: "The checkpoint's own attention.",
  sage: "Sage attention — H3's attention runs quantized. Faster, and lower peak VRAM. Needs ComfyUI-KJNodes and sageattention on an NVIDIA card.",
  kitchen: "Comfy Kitchen attention — core's own int8 kernel, nothing to install. Needs a ComfyUI whose build ships it.",
};

// Noun first, the way the cache pill reads ("cache off", "cache fast"): the
// pill has to say what it is a choice *about*, and a pill reading "kitchen"
// on its own says nothing at all to somebody who has not read the release
// notes. The backends keep their own names — they are what the packs and core
// call them, and searching for either finds the right page.
const ATTENTION_LABEL = {
  default: "attention default",
  sage: "attention sage",
  kitchen: "attention kitchen",
};

const BLOCK_CACHE_TITLE = {
  off: "Step caching is off.",
  safe: "FirstBlockCache, safest preset — fewest skipped steps.",
  fast: "FirstBlockCache, the pack's recommended preset.",
  aggressive: "FirstBlockCache, most skipping — fastest, furthest from a native render.",
  easy: "EasyCache — core's own step reuse, nothing to install. Cannot be combined with Spectrum.",
  tea: "TeaCache — skips transformer forwards on timestep similarity. Needs ComfyUI-MiniMaxH3-TeaCache.",
};

/**
 * Carry a workflow saved with the old `sage` switch onto the `attention` list.
 *
 * The switch became one option of a list, and the two cannot both be authority
 * or a node would say sage in one place and default in the other. So the switch
 * is read exactly once — the first time a node carrying it is drawn — moved onto
 * the list and cleared, and after that the list is the only thing that decides.
 * A node that never had it on passes straight through and nothing is written.
 *
 * Only while the list is still at its default: a workflow saved *after* the
 * rename has already answered this, including by answering "default".
 */
function adoptSage(widgets, set) {
  if (!widgets.sage?.value) return;
  if (String(widgets.attention?.value ?? "default") === "default") set("attention", "sage");
  set("sage", false);
}

/**
 * Read and write the real widgets, by name.
 *
 * The `{value, set}` pair every body hands this row and `turbo.js` — the one
 * thing all three node bodies had a character-for-character copy of, differing
 * only in what they called the map they were holding. Here rather than on a
 * shared base class, because that is all they had in common: their `commit`,
 * `destroy` and `adoptWeights` are three different jobs that happen to share
 * three names, and a base class holding one function would have bought that
 * function at the price of a hierarchy.
 *
 * No re-render on write. Everything that uses this commits — and renders — once
 * at the end rather than three times along the way; a body that wants the
 * redraw does it in its own `set`.
 *
 * @param {() => object} widgets  name -> real ComfyUI widget. A thunk because a
 *   body may be handed its widgets after it is built.
 * @param {() => void} [onChange] the node needs redrawing on the canvas
 */
export function widgetIO(widgets, onChange) {
  return {
    value: (name, fallback) => widgets()?.[name]?.value ?? fallback,
    set: (name, value) => {
      const widget = widgets()?.[name];
      if (!widget) return;
      widget.value = value;
      // Some of them — the seed's after-generate control — hang behaviour off
      // the callback rather than off the value, so it is not optional.
      widget.callback?.(value);
      onChange?.();
    },
  };
}

/**
 * @param {object} options
 * @param {object} options.widgets           name -> real ComfyUI widget
 * @param {(name, fallback) => any} options.value
 * @param {(name, value) => void} options.set write-through to the widget
 * @param {boolean} options.perSegment       true when there is more than one
 *                                           generation, which changes what the
 *                                           seed and step counts mean
 * @param {HTMLElement[]} [options.turbo]     the turbo switch's pills (see
 *   turbo.js), drawn with the accelerators because that is what it is — built
 *   by the caller because it needs the state, which this row otherwise doesn't
 * @param {HTMLElement[]} [options.trailing] appended after the accelerators —
 *   the weights pill, which belongs on this row because it is the other half of
 *   "how is this run" and nowhere else because it is not a sampler setting
 * @returns {HTMLElement}
 */
/**
 * One card's own seed, on the card's own editor.
 *
 * Not the sampler row: steps, cfg, the sampler and the accelerators describe
 * how the piece is run and there is one answer to that for the whole node. The
 * seed is the one number a card has business overriding, and the reason is
 * shooting a piece a pass at a time — retaking segment 2 under a single number
 * for the whole piece means rolling the number that made the take already kept
 * on segment 1, so the handle stops describing the piece. A take's seed is a
 * fact about the take.
 *
 * Unlit while the card inherits, which is every card until somebody rolls one
 * here — the row's rule everywhere else, and the honest reading: an inherited
 * seed is doing nothing to this card that it is not doing to all of them.
 *
 * @param {number|null} options.own    this card's seed, or null to inherit
 * @param {number} options.piece       the number on the node, which is what an
 *                                     inheriting card runs on
 * @param {(seed: number|null) => void} options.onChange  null clears the override
 * @param {number|null} [options.taken] the seed the card's take was made on,
 *   named in the tooltip because "which seed made this" is the whole reason
 *   the number is on screen at all
 */
export function segmentSeedPill({ own, piece, onChange, taken = null }) {
  const inherited = own === null || own === undefined;
  const shown = String(inherited ? piece : own);
  return el("div", { class: `mmc-pill mmc-pill-group${inherited ? "" : " on"}` }, [
    el("button", {
      class: "mmc-step mmc-seed-dice",
      title: t("Roll a seed for this card alone. The rest of the piece keeps the "
             + "number on the node."),
      onclick: () => onChange(Math.floor(Math.random() * 0xffffffff)),
    }, [icon("dice", 15)]),
    el("button", {
      class: "mmc-step mmc-seed-last",
      disabled: inherited,
      title: inherited
        ? t("This card runs on the piece's seed, which is what every card does "
          + "until you roll one here.")
        : t("Back to the piece's seed, {seed}.", { seed: piece }),
      onclick: () => { if (!inherited) onChange(null); },
    }, [icon("rewind", 15)]),
    el("input", {
      class: "mmc-seed-input",
      type: "text",
      value: shown,
      style: { width: `${Math.min(21, Math.max(5, shown.length + 1))}ch` },
      title: [
        inherited
          ? t("The piece's seed. Type a number to give this card one of its own.")
          : t("This card's own seed. The rest of the piece runs on {seed}.", { seed: piece }),
        taken === null ? null : t("Its take was made on {seed}.", { seed: taken }),
      ].filter(Boolean).join(" "),
      onchange: (event) => {
        const text = String(event.target.value).replace(/[^\d]/g, "");
        onChange(text === "" ? null : Number(text) || 0);
      },
      onpointerdown: (event) => event.stopPropagation(),
    }),
    // Which of the two numbers is in force. A readout, not a control — the
    // rewind beside it is how you go back — so it is a span and wears the
    // "fixed" seed mode's own quiet.
    el("span", {
      class: `mmc-seed-mode${inherited ? "" : " on"}`,
      text: inherited ? t("piece") : t("card"),
    }),
  ]);
}


export function samplingBar({ widgets, value, set, perSegment = false, turbo = [], trailing = [] }) {
  const pills = [];

  if (widgets.seed) {
    const control = value("control_after_generate", "fixed");
    // What the last queue actually ran on, offered back. Always drawn, beside
    // the dice it undoes — a control that appears only once it would do
    // something is a control nobody knows is there, which is the whole of what
    // was wrong with hiding it. Dimmed and inert until there is a seed to go
    // back to, and again once the seed already is that one.
    const last = lastSeed(widgets.seed);
    const reusable = last !== null && last !== Number(value("seed", 0));
    const seedText = String(value("seed", 0));
    pills.push(el("div", { class: "mmc-pill mmc-pill-group" }, [
      el("button", {
        class: "mmc-step mmc-seed-dice",
        title: t("Roll a new seed now"),
        onclick: () => set("seed", Math.floor(Math.random() * 0xffffffff)),
      }, [icon("dice", 15)]),
      el("button", {
        class: "mmc-step mmc-seed-last",
        // `.mmc-step:disabled` already dims it and refuses the cursor.
        disabled: !reusable,
        title: last === null
          ? t("Nothing queued yet — after a render this comes back to the seed it ran on")
          : t("Back to {seed}, the seed the last queue ran on", { seed: last }),
        onclick: () => { if (reusable) set("seed", last); },
      }, [icon("rewind", 15)]),
      el("input", {
        class: "mmc-seed-input",
        type: "text",
        value: seedText,
        // Sized to the digits it is holding. The field was 92px, which is a
        // guess that fits about eight digits — and seeds are ten, so the number
        // that identifies the render was the one thing on the row you could not
        // read. `ch` is exactly one digit in the mono face the stylesheet gives
        // this, so the pill breathes with its content instead: tight around a
        // hand-typed 7, wide enough for anything the dice can roll.
        style: { width: `${Math.min(21, Math.max(5, seedText.length + 1))}ch` },
        title: perSegment
          ? t("The piece's seed: every card runs on this number unless it was given "
            + "one of its own, which is set on the card.")
          : t("The seed of the one generation."),
        onchange: (event) => {
          const parsed = Number(String(event.target.value).replace(/[^\d]/g, "")) || 0;
          set("seed", parsed);
        },
        onpointerdown: (event) => event.stopPropagation(),
      }),
      // Quiet on "fixed", which is now the default and the state where nothing
      // happens; awake on the three that move the seed between queues. The same
      // rule the accelerator pills follow — what is doing something to your
      // render is what is lit — and it hands the digits back the emphasis they
      // were competing with.
      ...(widgets.control_after_generate ? [el("button", {
        class: `mmc-ghost mmc-seed-mode${control === "fixed" ? "" : " on"}`,
        title: t("What happens to the seed after each queue"),
        text: control,
        onclick: (event) => openChoicePopover(event.currentTarget, {
          title: t("After generate"),
          options: SEED_CONTROL,
          value: control,
          onPick: (picked) => set("control_after_generate", picked),
        }),
      })] : []),
    ]));
  }

  if (widgets.steps) {
    pills.push(stepperPill({
      value: Number(value("steps", 20)), min: 1, max: 200, step: 1,
      iconName: "steps", width: "42px",
      title: perSegment ? t("Denoising steps, per segment") : t("Denoising steps"),
      format: (n) => t("{n} steps", { n }),
      onChange: (next) => set("steps", next),
    }));
  }

  if (widgets.cfg) {
    pills.push(stepperPill({
      value: Number(value("cfg", 1)), min: 0, max: 30, step: 0.5, width: "52px",
      title: t("Classifier-free guidance. The distilled H3 checkpoints want 1.0, "
           + "and at 1.0 the negative is skipped entirely."),
      format: (n) => t("cfg {n}", { n: n.toFixed(1) }),
      onChange: (next) => set("cfg", next),
    }));
  }

  for (const [name, label] of [["sampler_name", "Sampler"], ["scheduler", "Scheduler"]]) {
    const widget = widgets[name];
    if (!widget) continue;
    const options = widget.options?.values || [];
    pills.push(el("button", {
      class: "mmc-pill",
      title: t(label),
      onclick: (event) => openChoicePopover(event.currentTarget, {
        title: t(label),
        options: typeof options === "function" ? options(widget) : options,
        value: widget.value,
        onPick: (picked) => set(name, picked),
      }),
    }, [el("span", { text: String(widget.value) })]));
  }

  // The flow shifts, H3's two clocks. Drawn as one compact stepper pair after
  // the scheduler because they are schedule too: the checkpoints' own values
  // by default, reset by the turbo switch to what the picked LoRA's card
  // names, and honest about which track a wrong value ruins first.
  //
  // Hidden unless Settings → Nodes asks for them: most rows never leave the
  // checkpoints' own schedule, and the widgets keep working underneath — the
  // turbo switch still writes them, loaded workflows still carry them. A value
  // *off* that schedule shows its pill whatever the setting says, the custom-CRF
  // rule: what is in force has to be visible, and hiding it is how a stale
  // turbo preset would quietly ruin every later render.
  const showShifts = uiSetting("show_shift_pills", false);
  // Whether this machine wants the controls most rows never touch. It hides,
  // it does not disable: a control that is *on* keeps its pill whatever this
  // says, so nothing can be switched on and out of sight at the same time.
  const advanced = uiSetting("advanced", false) === true;
  if (widgets.shift_video && (showShifts || Number(value("shift_video", 12)) !== 12)) {
    pills.push(stepperPill({
      value: Number(value("shift_video", 12)), min: 0.01, max: 100, step: 0.5, width: "48px",
      title: t("The video flow shift. 12 is the checkpoints' own schedule; a turbo LoRA's card may name another."),
      format: (n) => t("shift {n}", { n: +n.toFixed(2) }),
      onChange: (next) => set("shift_video", next),
    }));
  }
  if (widgets.shift_audio && (showShifts || Number(value("shift_audio", 3)) !== 3)) {
    pills.push(stepperPill({
      value: Number(value("shift_audio", 3)), min: 0.01, max: 100, step: 0.5, width: "48px",
      title: t("The audio flow shift. 3 is the checkpoints' own schedule. A wrong one distorts the soundtrack before it touches the picture."),
      format: (n) => t("audio {n}", { n: +n.toFixed(2) }),
      onChange: (next) => set("shift_audio", next),
    }));
  }

  // The accelerators. Off is the default and reads as off — an unlit pill —
  // because they are other people's nodes and a render with one on is not a
  // native render, which is worth being able to see at a glance. The turbo
  // switch leads them: it is the one that changes the most about the run.
  pills.push(...turbo);

  if (widgets.block_cache) {
    const options = widgets.block_cache.options?.values || [];
    const current = String(value("block_cache", "off"));
    pills.push(el("button", {
      class: `mmc-pill${current === "off" ? "" : " accel-on"}`,
      title: BLOCK_CACHE_TITLE[current] ? t(BLOCK_CACHE_TITLE[current]) : t("FirstBlockCache"),
      onclick: (event) => openChoicePopover(event.currentTarget, {
        title: t("Block cache"),
        options: typeof options === "function" ? options(widgets.block_cache) : options,
        value: current,
        onPick: (picked) => set("block_cache", picked),
      }),
    }, [el("span", { text: current === "off" ? t("cache off") : t("cache {preset}", { preset: current }) })]));
  }

  if (widgets.spectrum) {
    const on = Boolean(value("spectrum", false));
    pills.push(el("button", {
      class: `mmc-pill${on ? " accel-on" : ""}`,
      title: on
        ? t("Spectrum on — forecasting features across steps.")
        : t("Spectrum off. Needs ComfyUI-Spectrum-MiniMax-H3 when switched on."),
      onclick: () => set("spectrum", !on),
    }, [el("span", { text: on ? t("spectrum") : t("spectrum off") })]));

    // Only worth a control when it is doing something; the blend is ignored
    // outright while Spectrum is off.
    if (on && widgets.spectrum_blend) {
      pills.push(stepperPill({
        value: Number(value("spectrum_blend", 0.5)), min: 0, max: 1, step: 0.05, width: "52px",
        title: t("Spectrum's video spectral share — higher is faster and further from a native render"),
        format: (n) => t("blend {n}", { n: n.toFixed(2) }),
        onChange: (next) => set("spectrum_blend", next),
      }));
    }
  }

  // Last of the accelerators, and the two that do not change which steps run —
  // one changes what an attention call costs and the other what the MLP peaks
  // at, so they sit with them but rule nothing else out.
  if (widgets.attention) {
    adoptSage(widgets, set);
    const options = widgets.attention.options?.values || [];
    const current = String(value("attention", "default"));
    pills.push(el("button", {
      class: `mmc-pill${current === "default" ? "" : " accel-on"}`,
      title: t(ATTENTION_TITLE[current] || ATTENTION_TITLE.default),
      onclick: (event) => openChoicePopover(event.currentTarget, {
        title: t("Attention"),
        options: typeof options === "function" ? options(widgets.attention) : options,
        value: current,
        onPick: (picked) => set("attention", picked),
      }),
    }, [el("span", { text: t(ATTENTION_LABEL[current] || ATTENTION_LABEL.default) })]));
  }

  // The last two are named for what you get rather than for how they are
  // built. "Chunk FFN" and "fp16 accumulation" are what the packs call them and
  // are in the tooltips, where somebody looking for them will find them; on the
  // row they are the only pills that would have asked you to know what a
  // feed-forward is before you could tell whether you wanted one.
  if (widgets.chunk_ffn && (advanced || Boolean(value("chunk_ffn", false)))) {
    const on = Boolean(value("chunk_ffn", false));
    pills.push(el("button", {
      class: `mmc-pill${on ? " accel-on" : ""}`,
      title: on
        ? t("Low VRAM on — H3's feed-forward runs in chunks, so the peak is lower. The frames are the same ones: chunking is a rearrangement, not a trade. Needs ComfyUI-KJNodes.")
        : t("Low VRAM — run H3's feed-forward in chunks (KJNodes' Chunk FFN). Lowers the peak a render has to fit in, and changes nothing about the frames. Needs ComfyUI-KJNodes."),
      onclick: () => set("chunk_ffn", !on),
    }, [el("span", { text: on ? t("low vram") : t("low vram off") })]));
  }

  if (widgets.fp16_accumulation && (advanced || Boolean(value("fp16_accumulation", false)))) {
    const on = Boolean(value("fp16_accumulation", false));
    pills.push(el("button", {
      class: `mmc-pill${on ? " accel-on" : ""}`,
      title: on
        ? t("Fast math on — cuBLAS accumulates fp16 matmuls in fp16 while this model runs (KJNodes' fp16 accumulation). Only fp16 matmuls: a bf16 or quantized H3 has none, and nothing changes there.")
        : t("Fast math — let cuBLAS accumulate fp16 matmuls in fp16 while this model runs (KJNodes' fp16 accumulation). Faster where the card supports it, at some precision. Only reaches a genuinely fp16 model: the released H3 checkpoints run bf16, and their quantized layers go through comfy-kitchen's own kernels rather than cuBLAS, so on those this does nothing at all. Needs ComfyUI-KJNodes and torch 2.7 or newer."),
      onclick: () => set("fp16_accumulation", !on),
    }, [el("span", { text: on ? t("fast math") : t("fast math off") })]));
  }

  return el("div", { class: "mmc-pills" }, [...pills, ...trailing]);
}

File: creator/families/ltx25/sampling.py
"""LTX 2.5's sampler row — the defaults, and the nodes each one belongs to.

The row is a different shape from H3's, which is the reason `families/manifest.py`
describes controls instead of shipping a bag of values. H3 emits a `KSampler`
and its row is that node's inputs. LTX emits three nodes:

    ManualSigmas(sigmas) | LTXVScheduler(steps, ..., latent)               -> SIGMAS
    KSamplerSelect(sampler_name)                                           -> SAMPLER
    LTXVDualCFGGuider(model, positive, negative, video_cfg, audio_cfg)     -> GUIDER
                                                    -> SamplerCustomAdvanced

so there is no `scheduler` combo at all, and one `cfg` becomes two — the packed
AV latent takes a separate scale per modality.

**`schedule` is which of the two SIGMAS nodes the render emits, and it is the
field the rest of the row hangs off.** The two transformers in the `dit` slot
are sampled differently rather than merely at different settings:

- `distilled` feeds `declare.DISTILLED_SIGMAS` through `ManualSigmas` and emits
  no `ModelSamplingLTXV` at all. Neither number is ours and neither is a
  default — the curve is what the checkpoint was distilled against, so `steps`,
  `max_shift`, `base_shift`, `stretch` and `terminal` describe nothing on this
  route and are not read. This is what both of Lightricks' 2.5 workflows and
  ComfyUI's own LTX 2.5 templates do.
- `scheduler` is `LTXVScheduler` with the shift pair and the terminal, paired
  with the `ModelSamplingLTXV` patch that reads the same curve — the recipe LTX
  2.3 shipped, and what the full `dev` transformer wants, at ~20 steps and
  3.0/7.0 rather than 8 at 1/1.

Distilled is the default because the distilled transformer is what a piece
loads unless someone picks otherwise, and because the trained curve is the one
thing about this family that a wrong guess costs the whole render — see
`declare.DISTILLED_SIGMAS` for the two curves side by side. The bounds below
stay wide enough for both routes.

`max_shift`, `base_shift`, `stretch` and `terminal` are `LTXVScheduler`'s own
numbers, unchanged: they describe the sigma curve the architecture was trained
against, not a taste.

**The three guidance fields are a different kind of control** and are declared
in their own group for it. `stg_scale`, `stg_blocks` and `modality_scale` do not
describe the schedule: each switches on an *extra forward pass per step*, and
the pass they add is the same size as the one the render was already making. So
they are off by default at the values core calls off — 0 for STG, 1.0 for
modality guidance — and the pills that write them say what they cost rather than
leaving a user to find out from the step timer. See `render._guided`.
"""

from dataclasses import dataclass

from ... import accel
from .. import row as base_row
from . import declare

#: Which SIGMAS node the render emits. See the module docstring.
SCHEDULES = ("distilled", "scheduler")


@dataclass(frozen=True)
class Sampling:
    """The row, however the piece came by it.

    The same role H3's `sampling.Sampling` plays and deliberately not the same
    fields: `sampler_name` is the only name the two share, because the two
    families' rows describe different nodes. `steps` is here under H3's name for
    the one reason a shared name is ever worth it — the render loop passes this
    object through unread, but `core/emit.py`'s progress and the accelerator
    planner both count steps, and a family that spelled it otherwise would be
    inventing a difference.
    """

    seed: int = 0
    schedule: str = SCHEDULES[0]
    steps: int = 8
    video_cfg: float = 1.0
    audio_cfg: float = 1.0
    sampler_name: str = declare.DISTILLED_SAMPLER
    max_shift: float = 2.05
    base_shift: float = 0.95
    stretch: bool = True
    terminal: float = 0.1
    #: Taste guidance, off at core's own off values. See the module docstring.
    stg_scale: float = 0.0
    stg_blocks: str = "29"
    modality_scale: float = 1.0

    @property
    def manual(self):
        """Whether the render samples on the checkpoint's trained curve.

        The one thing outside this module that has to ask: `render._schedule`
        emits a different SIGMAS node either way, and on this route it emits no
        `ModelSamplingLTXV` either.
        """
        return self.schedule == "distilled"

    @property
    def stg(self):
        """Whether STG would do anything. Both halves have to say yes: a scale
        of 0 short-circuits inside core's post-CFG hook, and so does an empty
        block set, so either one alone means an extra node in the graph that
        costs a model clone and changes no frame."""
        return self.stg_scale != 0.0 and any(c.isdigit() for c in self.stg_blocks)

    @property
    def modality(self):
        """Whether modality guidance would do anything. 1.0 is core's own
        "no extra pass", not a low setting."""
        return self.modality_scale != 1.0


DEFAULTS = {
    # The trained curve, which is what the file in the `dit` slot is unless a
    # user went and picked the dev transformer. See the module docstring.
    "schedule": SCHEDULES[0],
    # `LTXVScheduler`'s step count, and the `scheduler` route's alone — the
    # distilled curve carries its own eight and reads nothing from here. 8
    # rather than the node's own 20 so that switching routes lands on the
    # distilled row's shape rather than the dev transformer's.
    "steps": 8,
    # One scale per modality of the packed AV latent. Both 1.0 is the distilled
    # checkpoint's; the dev transformer's are the guider's own 3.0 / 7.0.
    "video_cfg": 1.0,
    "audio_cfg": 1.0,
    # The sigma curve. `ModelSamplingLTXV` takes the same shift pair and must be
    # given the same numbers — the model patch and the schedule are two readings
    # of one curve, and letting them disagree is a silent quality bug.
    "max_shift": 2.05,
    "base_shift": 0.95,
    "stretch": True,
    "terminal": 0.1,
    # `KSamplerSelect`'s list, which is core's to declare — the manifest's combo
    # carries no options for the same reason H3's does not. Ancestral, which is
    # what both official workflows pick in both stages: the noise an ancestral
    # step adds back is part of what eight steps were distilled against, and
    # plain euler over the same curve comes out flat.
    "sampler_name": declare.DISTILLED_SAMPLER,
    # Taste guidance, off. `LTXVSpatioTemporalGuidance` returns the unmodified
    # prediction at 0 and `LTXVModalityGuidance` at 1.0 — these are core's own
    # off values rather than a timid setting, which is why the pills read them
    # through the manifest's `off` rather than through a range.
    "stg_scale": 0.0,
    # `LTXVSpatioTemporalGuidance`'s own default block, kept: which blocks to
    # degrade is a property of the trained transformer, and 29 is what
    # Lightricks' node ships pointing at.
    "stg_blocks": "29",
    "modality_scale": 1.0,
}


# What each field has to be. The same refusal-by-name H3's row makes: a
# hand-edited `"steps": "eight"` should say so before a loader is built.
_WHOLE = ("steps",)
_NUMBER = ("video_cfg", "audio_cfg", "max_shift", "base_shift", "terminal",
           "stg_scale", "modality_scale")
_FLAG = ("stretch",)
# The one field with a closed list. A misspelt route would otherwise pass as a
# name and sample on the wrong curve, which is the failure this whole field
# exists to make impossible.
_CHOICE = {"schedule": SCHEDULES}


class SamplingError(ValueError):
    """A `sampling` block this family will not run."""


# The floor a field's own node imposes, where it is not zero. Modality guidance
# is the whole list: `LTXVModalityGuidance` takes 1.0 as "off" and its formula
# scales by `modality_scale - 1`, so a value below it pushes the prediction away
# from the coupled one — the opposite of the control's name.
_FLOOR = {"stg_scale": 0.0, "modality_scale": 1.0}


def _blocks(value):
    """`stg_blocks`, checked. -> the string as written.

    Core parses it with `re.findall(r"\\d+")`, so what is legal is "any text
    with some numbers in it" — and an empty one is legal too, meaning STG is
    off however high its scale is. What is refused is text with no number in it
    at all, which is the typo case: it would sample at full cost with the
    perturbation switched off, and say nothing.

    A `custom` field on the row below rather than a kind of its own, because
    nothing else in this pack has an opinion like it.
    """
    if not isinstance(value, str):
        raise SamplingError("sampling.stg_blocks must be a list of block "
                            "numbers, written as text")
    if value.strip() and not any(c.isdigit() for c in value):
        raise SamplingError(
            f"sampling.stg_blocks is {value!r}, which names no block. Give it "
            f"transformer block numbers separated by commas, or leave it empty "
            f"to switch spatio-temporal guidance off.")
    return value


ROW = base_row.Row(DEFAULTS, error=SamplingError, whole=_WHOLE, number=_NUMBER,
                   flag=_FLAG, choice=_CHOICE, floors=_FLOOR,
                   custom={"stg_blocks": _blocks})


def resolve(data, widgets):
    """-> `(Sampling, accel.Settings)` for this queue.

    The blob is the whole of the row. There is no widget fallback the way H3 has
    one, and there could not be: the node's thirteen widget slots are H3's row,
    frozen in that order forever, and none of them is `video_cfg`. A field the
    blob does not carry falls back to `DEFAULTS` — the distilled checkpoint's
    numbers — which is what an LTX piece written by any frontend older than the
    field should run at.

    The seed is the exception and still comes off the widget, for the reason it
    does on H3: `control_after_generate` is the frontend's own linked control
    and there is nothing for a JSON field to be.

    The accelerators come back off. Every one of them is H3's — the block
    caches are keyed to its DiT, Spectrum ships a pack named for it, and the two
    attention swaps and the FFN chunker are KJNodes patches written against its
    blocks. None of that has been tried against LTX, and quietly applying a
    model patch nobody has verified is how a render comes out subtly wrong with
    nothing in the log. Off is the honest answer until someone measures.
    """
    stored = ROW.stored(data)

    def pick(name):
        return stored[name] if name in stored else DEFAULTS[name]

    return (
        Sampling(
            seed=int(widgets.get("seed", 0)),
            schedule=pick("schedule"), steps=pick("steps"),
            video_cfg=pick("video_cfg"), audio_cfg=pick("audio_cfg"),
            sampler_name=pick("sampler_name"),
            max_shift=pick("max_shift"), base_shift=pick("base_shift"),
            stretch=pick("stretch"), terminal=pick("terminal"),
            stg_scale=pick("stg_scale"), stg_blocks=pick("stg_blocks"),
            modality_scale=pick("modality_scale"),
        ),
        accel.Settings(),
    )
